### PR TITLE
fix(lib/trie): Check for root in EncodeAndHash

### PR DESCRIPTION
**Pull request by kishansagathiya**
## Changes

- In EncodeAndHash, If a node is root, we return blake2b hash of its encoding as hash
regardless of whether length of encoding is smaller or greater than 32

## Tests

```

```

## Issues

- #2357

## Primary Reviewer

- @qdm12

### CODE REVIEW

File: internal/trie/node/hash_test.go [hunk 1]
```diff
@@ -119,7 +119,7 @@ func Test_Branch_EncodeAndHash(t *testing.T) {
 		t.Run(name, func(t *testing.T) {
 			t.Parallel()
 
-			encoding, hash, err := testCase.branch.EncodeAndHash()
+			encoding, hash, err := testCase.branch.EncodeAndHash(false)
```
Comment (qdm12) [inline internal/trie/node/hash_test.go:122]: please add `isRoot` as a field to the test case. And also add a test case with `isRoot: true`.

File: internal/trie/node/hash_test.go [hunk 2]
```diff
@@ -241,7 +241,7 @@ func Test_Leaf_EncodeAndHash(t *testing.T) {
 		t.Run(name, func(t *testing.T) {
 			t.Parallel()
 
-			encoding, hash, err := testCase.leaf.EncodeAndHash()
+			encoding, hash, err := testCase.leaf.EncodeAndHash(false)
```
Comment (qdm12) [inline internal/trie/node/hash_test.go:244]: please add `isRoot` as a field to the test case. And also add a test case with `isRoot: true`.

File: lib/trie/lookup.go [hunk 3]
```diff
@@ -18,24 +18,24 @@ type recorder interface {
 
 // findAndRecord search for a desired key recording all the nodes in the path including the desired node
 func findAndRecord(t *Trie, key []byte, recorder recorder) error {
-	return find(t.root, key, recorder)
+	return find(t.root, key, recorder, true)
```
Comment (qdm12) [inline lib/trie/lookup.go:21]: nit for extreme readability

```suggestion
	const isRoot = true
	return find(t.root, key, recorder, isRoot)
```

File: lib/trie/lookup.go [hunk 4]
```diff
@@ -18,24 +18,24 @@ type recorder interface {
...
 }
 
-func find(parent Node, key []byte, recorder recorder) error {
-	enc, hash, err := parent.EncodeAndHash()
+func find(current Node, key []byte, recorder recorder, isCurrentRoot bool) error {
```
Comment (qdm12) [inline lib/trie/lookup.go:24]: why not leave it as `parent`? As in it's `parent` elsewhere mostly.
Although `current` works as well I guess.

File: lib/trie/lookup.go [hunk 5]
```diff
@@ -49,5 +49,5 @@ func find(parent Node, key []byte, recorder recorder) error {
 		return nil
 	}
 
-	return find(b.Children[key[length]], key[length+1:], recorder)
+	return find(b.Children[key[length]], key[length+1:], recorder, false)
```
Comment (qdm12) [inline lib/trie/lookup.go:52]:
```suggestion
	const isChildRoot = false
	return find(b.Children[key[length]], key[length+1:], recorder, isChildRoot)
```

**Review by qdm12 — APPROVED** [review]
Nice, just a few nits.

Also, would it be possible to instead change the database store/fetch code to be able to understand smaller than 32 bytes values instead of hash for the root node? I guess not, just asking this out of curiosity.

File: lib/trie/lookup.go (hunk 4 above)
Comment (kishansagathiya) [reply lib/trie/lookup.go:24]: parent seems rather misleading. parent suggest that there is some other node that is more interesting for this function

Comment (qdm12) [reply lib/trie/lookup.go:24]: We are finding a node which can either be the parent itself or one of its children. Why is it misleading?

In some function like `insert` we return a new node given an existing node and a node to insert; how should we call the returned node? I'm happy to do a batch rename if you have better naming. For now I was using `parent` as the argument and `newParent` as the returned node, but maybe `currentNode` and `newNode` would be better? :thinking:

**Comment by kishansagathiya** [comment]
> Also, would it be possible to instead change the database store/fetch code to be able to understand smaller than 32 bytes values instead of hash for the root node?

This could be possible. I prefer the current approach however. I prefer function's behaviour to be obvious based on its signature. If I edit `key` inside say db.Put(key, value), that is not an obvious behaviour. A behaviour like that could lead to bugs!

**Comment by qdm12** [comment]
>  If I edit key inside say db.Put(key, value), that is not an obvious behaviour. A behaviour like that could lead to bugs!

Definitely. But can't we write keys as byte slices (not fixed size arrays) to database? That way there might not be any custom code for the root, and we can know if it's a node value or the node encoding hash digest depending on the database retrieved byte slice length? :thinking: Maybe I'm wrong though..

File: lib/trie/lookup.go (hunk 5 above)
Comment (EclesioMeloJunior) [reply lib/trie/lookup.go:52]: the `isChildRoot` variable does not make much sense, the root node already has been processed and the other nodes are not the root, so it is explicit that for other `find` calls the `false` argument means the nodes are not the root

**Review by EclesioMeloJunior — APPROVED** [review]

File: lib/trie/database.go [hunk 6]
```diff
@@ -97,7 +97,7 @@ func (t *Trie) loadFromProof(rawProof [][]byte, rootHash []byte) error {
 		decodedNode.SetDirty(dirty)
 		decodedNode.SetEncodingAndHash(rawNode, nil)
 
-		_, hash, err := decodedNode.EncodeAndHash()
+		_, hash, err := decodedNode.EncodeAndHash(decodedNode == t.root)
```
Comment (noot) [inline lib/trie/database.go:100]: `decodedNode == t.root` is always false here (which is fine, since it's the same functionality as before)

Comment (noot) [reply lib/trie/database.go:100]: but perhaps just replace the case with `false`?

File: lib/trie/lookup.go (hunk 4 above)
Comment (kishansagathiya) [reply lib/trie/lookup.go]: changed back to parent

**Pull request merged**